oracle/src/oraclewizard.cpp

```diff
@@ -529,3 +544,203 @@ bool SaveSetsPage::validatePage()
 
     return true;
 }
+
+LoadTokensPage::LoadTokensPage(QWidget *parent)
+    : OracleWizardPage(parent), nam(0)
+{
+    urlLabel = new QLabel(this);
+    urlLineEdit = new QLineEdit(this);
+
+    progressLabel = new QLabel(this);
+    progressBar = new QProgressBar(this);
+
+    urlButton = new QPushButton(this);
+    connect(urlButton, SIGNAL(clicked()), this, SLOT(actRestoreDefaultUrl()));
+
+    QGridLayout *layout = new QGridLayout(this);
+    layout->addWidget(urlLabel, 0, 0);
+    layout->addWidget(urlLineEdit, 0, 1);
+    layout->addWidget(urlButton, 1, 1, Qt::AlignRight);
+    layout->addWidget(progressLabel, 2, 0);
+    layout->addWidget(progressBar, 2, 1);
+
+    setLayout(layout);
+}
+
+void LoadTokensPage::initializePage()
+{
+    urlLineEdit->setText(wizard()->settings->value("tokensurl", TOKENS_URL).toString());
+
+    progressLabel->hide();
+    progressBar->hide();
+}
+
+void LoadTokensPage::retranslateUi()
+{
+    setTitle(tr("Tokens source selection"));
+    setSubTitle(tr("Please specify a source for the list of tokens. "
+                   "You can specify an url address that will be download or "
+                   "use an existing file from your computer."));
+
+    urlLabel->setText(tr("Download url:"));
+    urlButton->setText(tr("Restore default url"));
+}
+
+void LoadTokensPage::actRestoreDefaultUrl()
+{
+    urlLineEdit->setText(TOKENS_URL);
+}
+
+bool LoadTokensPage::validatePage()
+{
+    // once the import is finished, we call next(); skip validation
+    if(wizard()->hasTokensData())
+        return true;
+
+    QUrl url = QUrl::fromUserInput(urlLineEdit->text());
+    if(!url.isValid())
+    {
+        QMessageBox::critical(this, tr("Error"), tr("The provided url is not valid."));
+        return false;
+    }
+
+    progressLabel->setText(tr("Downloading (0MB)"));
+    // show an infinite progressbar
+    progressBar->setMaximum(0);
+    progressBar->setMinimum(0);
+    progressBar->setValue(0);
+    progressLabel->show();
+    progressBar->show();
+
+    wizard()->disableButtons();
+    setEnabled(false);
+
+    if(!nam)
+        nam = new QNetworkAccessManager(this);
+    QNetworkReply *reply = nam->get(QNetworkRequest(url));
+
+    connect(reply, SIGNAL(finished()), this, SLOT(actDownloadFinishedTokensFile()));
+    connect(reply, SIGNAL(downloadProgress(qint64, qint64)), this, SLOT(actDownloadProgressTokensFile(qint64, qint64)));
```

Comment:
error() is emitted when an error occurs. finished() is emitted when the request ends, both if it succeeded or it failed.